### ISSUE TITLE
[AP proposal] Toward a more unified UX: BFF 💕 

### ISSUE THREAD

**Issue by andreaTP** [bf2fc6cc711aee1a0c2a/architecture#91]
## Problem statement

The integration of the Enterprise option of RHOSAK in UI, SDKs and CLI has shown that there is a margin for improvement in how we expose and consume our services APIs.

More specifically:
- the current services are offering the "bare functionality" (as they should) without providing many helpers
- the data aggregation logic is scattered, duplicated, and possibly even diverging in various codebases (especially UI and CLI)
- the integration of the various services is done in multiple places
- to achieve a desired functionality through the REST APIs the end user is exposed to the entire complexity of the underlying infrastructure

One of the patterns applied to improve when those issues arise is [BFF](https://samnewman.io/patterns/architectural/bff/) 💕 .

# Backend For Frontend

PROS:

- aggregate business logic to provide functionalities to users
- exposes endpoints for direct consumption from UX
- can be stateless (less maintenance, easy scaling up/down)
- UI and CLI need to know much less about the underlying infrastructure and can be thinner
- testing of the user-facing components can be done by mocking a single service instead of multiple APIs
- centralized patterns can be applied to a single component, such as rate-limiting etc.

CONS:

- an additional service to operate in the backend
- maintenance of a shared codebase for multiple teams (UI and DevExp)

Open questions:

- how do the current REST APIs fit into the desired User Experience?
- should we keep exposing the "bare functionality" or should it become a "technical detail" for the end user?

the idea has been conceived by speaking with @riccardo-forina and shared with the DevExp team( @dimakis ); we are convinced, at this point, that the advantages are outweighing the disadvantages but we are happy to receive early feedback before opening a formal AP.

**Comment by EricWittmann** [comment]
Interesting approach, for sure.  I'm sympathetic to it.  Can you give some concrete strawman examples of the functionality this BFF would provide?

**Comment by machi1990** [comment]
> cc. @EricWittmann @MikeEdgar @machi1990 @tombentley @lburgazzoli @emmanuelbernard @miguelsorianod @shawkins for feedback, and feel free to extend the visibility to anyone that might be interested in chipping into the conversation!

Adding @pb82 as well.

**Comment by tombentley** [comment]
Stupid question: What are the backends you envisage? 
Generally it would really help (me at least) if you could give concrete examples for each of the current problems you identify.

**Comment by andreaTP** [comment]
I'll coordinate with @riccardo-forina and @dimakis to provide a concrete example based on the recent experience of Enterprise 👍

**Comment by lburgazzoli** [comment]
I'm very sympathetic to the idea as the topics pops up every now and then in the RHOC team as it is very complicated i.e. for the UI to deal with all the low building block that often cross the single service boundaries.

**Comment by EricWittmann** [comment]
> I'm very sympathetic to the idea as the topics pops up every now and then in the RHOC team as it is very complicated i.e. for the UI to deal with all the low building block that often cross the single service boundaries.

+1 - I think this problem is likely to get worse as we continue to try and integrate services more tightly in the UI.

Do you imagine these BFF APIs to be available to customers as well?

**Comment by lburgazzoli** [comment]
> > I'm very sympathetic to the idea as the topics pops up every now and then in the RHOC team as it is very complicated i.e. for the UI to deal with all the low building block that often cross the single service boundaries.
> 
> +1 - I think this problem is likely to get worse as we continue to try and integrate services more tightly in the UI.
> 
> Do you imagine these BFF APIs to be available to customers as well?

If done right, they could be a differentiator so I'd say at some point the could be made available to everyone.

**Comment by riccardo-forina** [comment]
I actually have a different view for this. I see the existing APIs as the low level, non-MAS specific ones available to all upstream users to use as it works best for them. The BFF on the other hand should be heavily MAS specific as it needs to deal with concerns that us and us only (e.g. all the quota and billing bits)

**Comment by dimakis** [comment]
> Stupid question: What are the backends you envisage?
> Generally it would really help (me at least) if you could give concrete examples for each of the current problems you identify.

Taking an example from Hybrid (Enterprise) the displaying of Hybrid cluster names and that cluster's Id during the CLI's `list` or during `create kafka` commands. This is currently two separate calls, one to KFM for cluster details and the other to OCM just to fetch the cluster name. To grab cluster capacity a further call to KFM is needed per individual cluster.  

Furthermore, with the current approach, this logic would be duplicated in the CLI and UI. Even though at present, it is not really possible to do this well in the UI, as OCM doesn't provide a JavaScript/TypeScript SDK. A fetch to URL or rolling out our own generated SDK of OCM are the kind of options that have been discussed as a result. 

There are further bits of duplicated logic when querying AMS for the user's quota etc. too.

**Comment by andreaTP** [comment]
> What are the backends you envisage?

@tombentley what @dimakis mentioned above ☝️ and also from @riccardo-forina experience.
Creating a Kafka Instance when Enterprise is active roughly resemble to:

- getting information for the current account (multiple calls to get quotas, entitlements, current usage etc. etc.)
- gettting cloud providers/regions and more available options from fleet manager
- obtain information regarding the currently deployed instances of Kafka
- combine all the information above to know the available options for the user

Looking at the code, you can notice that all the merging of the data to understand quotas/availability/possible options is done both in CLI and UI.
Also, just to show the "Create Kafka" screen (or show the available options in CLI interactive mode), we are approx in between 3 to 10+ API calls and this is starting to impact performance too.

Having a server side component that maps one UI/CLI concern to one API call would be extremely beneficial and can be subsequently further optimized with caching and more.

> Do you imagine these BFF APIs to be available to customers as well?

@EricWittmann this is a super interesting discussion, but I would feel more comfortable in having it after we decide to have this component in place for internal usage first.
Realistically speaking, covering the needs of CLI/UI *and customers* with a single API seems like a stretch and highly subject to premature over engineering, but I might be wrong ... I propose to re-evaluate this point with the BU after we have a first implementation in place.

**Comment by tombentley** [comment]
Is there one BFF for each service?

**Comment by tombentley** [comment]
(BTW thanks for the clarifications so far!)

**Comment by andreaTP** [comment]
> Is there one BFF for each service?

NO, I think this point has been made clear by the visuals added at the top.

**Comment by emmanuelbernard** [comment]
Thanks for this proposition. I have the following questions I would like to see addressed:

- the BBF will call into the data plane admin API: how is this going to work with Enterprise Plan with VPC isolation (I'm assuming the customer would get access to this VPC from where it is calling the CLI).
- because it becomes a public API used by the UI and the CLI, how can we make it *not* a supported API. I'm concerned about the proliferation of APIs though I'm sympathetic to the BBF pattern.
- What is the expectation, one BBF for all key exposed services (Service Registry, Connectors, Kafka etc) or one per such service? The drawing only focus on Kafka usage.
- what makes it not possible for our API to address the UI and CLI needs (esp why isn't kas-fleetmanager API not shielding the quota management aspect). The reasoning is that if we need to do it why wouldn't a user of the API need it too?
- is the CLI and UI the same calls? The classic BBF patterns tends to promote one BBF per front end (one of CLI, one for UI, etc).

**Comment by emmanuelbernard** [comment]
By suppored API I mean we never break it which seems a bit counter to the idea of the BBF flexibility you are looking for.

**Comment by emmanuelbernard** [comment]
@andreaTP let's discuss it at the arch call tomorrow
